src/business-logic/services/inews-gateway-connection-implementation.ts

```diff
@@ -0,0 +1,128 @@
+    // For now: Simulate an async operation, e.g., network request
+    await new Promise(resolve => setTimeout(resolve, 1000))
+  }
+
+  public async listen(_deviceId: string, _callback: (data: unknown) => void): Promise<void> {
+    // For now: Simulate an async operation, e.g., network request
+    await new Promise(resolve => setTimeout(resolve, 1000))
+  }
+
+  public async connect(): Promise<boolean> {
+    return new Promise<boolean>((resolve, reject) => {
+      if (this.isConnectingOrDisconnecting) {
+        return reject(new Error('Device is already connecting or disconnecting.'))
+      }
+
+      try {
+        this.client = new WebSocket(this.INEWS_GATEWAY_HOST)
+        this.initializeClientEvents()
+        INewsGatewayDeviceConnection.device.isConnected = true
+        resolve(true)
+  
+      } catch (error) {
+        INewsGatewayDeviceConnection.device.isConnected = false
+        reject(new Error('WebSocket connection failed: ' + error))
+        return false
+      } finally {
+        this.isConnectingOrDisconnecting = false
+      }
+    })
+  }
+
+  private initializeClientEvents(): boolean {
+    this.client.on('error', (error) => {
+      this.logger.error('WebSocket error:', error)
+      INewsGatewayDeviceConnection.device.isConnected = false
+    })
+  
+    this.client.on('open', () => {
+      this.heartbeat()
+    })
+
+    this.client.on('ping', () => {
+      this.heartbeat()
+    })
+
+    this.client.on('close', (code, reason) => {
+      if (this.pingTimeout !== null) {
+        clearTimeout(this.pingTimeout)
+      }
+
+      if (INewsGatewayDeviceConnection.device.isConnected) {
+        INewsGatewayDeviceConnection.device.isConnected = false
+      }
+      this.logger.info(`WebSocket closed: Code ${code}, Reason: ${reason}`)
+
+      // reconnection logic
+    })
+
+    return INewsGatewayDeviceConnection.device.isConnected
+  }
+
+  private heartbeat(): void {
+    if (this.pingTimeout !== null) {
+      clearTimeout(this.pingTimeout)
+    }
+
+    this.pingTimeout = setTimeout(() => {
+      this.terminate()
+    }, 30000 + 1000)
+  }
+
+  private terminate(): void {
+    if (this.client.readyState === WebSocket.OPEN) {
+      this.client.terminate()
+    }  
+  }
+
+  public async disconnect(): Promise<boolean> {
+    return new Promise<boolean>((_resolve, reject) => { 
+      if (this.isConnectingOrDisconnecting) {
+        return reject(new Error('Device is already connecting or disconnecting.'))
+      }
+      this.isConnectingOrDisconnecting = true
+
+      try {
+        if (this.client.readyState === WebSocket.OPEN) {
+          this.client.close()
+        }
+    
+        if (this.pingTimeout !== null) {
+          clearTimeout(this.pingTimeout)
+        }      
+        INewsGatewayDeviceConnection.device.isConnected = false
+        return true
+      } catch (error) {
+        INewsGatewayDeviceConnection.device.isConnected = true
+        return false
```

Comment:
Kind of the same as already discussed a bit about what to return for Promise based methods, but wouldn't it be more useful for the caller to get a rejected promise with the caught error, rather than a successful call that says that it wasn't successful?